src/Service/Resolvers/MsSqlQueryBuilder.cs

```diff
@@ -110,28 +113,69 @@ public string Build(SqlExecuteStructure structure)
+            string tableName = $"{QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)}";
+
+            // Predicates by vritue of PK.
+            string pkPredicates = JoinPredicateStrings(Build(structure.Predicates));
+
+            // Predicates by virtue of PK + database policy.
+            string updatePredicates = JoinPredicateStrings(pkPredicates, structure.DbPolicyPredicatesForOperations[Config.Operation.Update]);
+
+            string updateOperations = Build(structure.UpdateOperations, ", ");
+            string outputColumns = MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted);
+            string queryToGetCountOfRecordWithPK = $"SELECT COUNT(*) as {COUNT_ROWS_WITH_GIVEN_PK} FROM {tableName} WHERE {pkPredicates}";
+
+            // Query to initiate transaction and get number of records with given PK.
+            string prefixQuery = $"SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;" +
+                $"BEGIN TRANSACTION;" +
+                $"DECLARE @ROWS_TO_UPDATE int;" +
+                $"SET @ROWS_TO_UPDATE = ({queryToGetCountOfRecordWithPK}); " +
+                $"{queryToGetCountOfRecordWithPK};";
+
+            // Final query to be executed for the given PUT/PATCH operation.
+            StringBuilder upsertQuery = new(prefixQuery);
+
+            // Query to update record (if there exists one for given PK).
+            StringBuilder updateQuery = new(
+                $"IF @ROWS_TO_UPDATE = 1" +
+                $"UPDATE {tableName} WITH(UPDLOCK) " +
+                $"SET {updateOperations} " +
+                $"OUTPUT {outputColumns} " +
+                $"WHERE {updatePredicates};");
+
+            // Append the update query to upsert query.
+            upsertQuery.Append(updateQuery);
+
+            if (!structure.IsFallbackToUpdate)
             {
-                return $"SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;BEGIN TRANSACTION; UPDATE {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} " +
-                    $"WITH(UPDLOCK) SET {Build(structure.UpdateOperations, ", ")} " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"WHERE {predicates} " +
-                    $"IF @@ROWCOUNT = 0 " +
-                    $"BEGIN; " +
-                    $"INSERT INTO {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} ({Build(structure.InsertColumns)}) " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"VALUES ({string.Join(", ", structure.Values)}) " +
-                    $"END; COMMIT TRANSACTION";
+                // Append the conditional to check if the insert query is to be executed or not.
+                // Insert is only attempted when no record exists corresponding to given PK.
```

Comment:
>// Insert is only attempted when no record exists corresponding to given PK. [](http://example.com/codeflow?start=16&length=76)

I would also mention whether db policy result from update plays a role in whether to proceed with insert.

Reply:
Since it is mentioned that insert is only attempted when no record exists for given PK, why do we need to mention db policy here?

Reply:
"when no record exists for given PK" -> could this ever be due to  dbpolicy constraints resulting in zero rows returned from update?

Reply:
No